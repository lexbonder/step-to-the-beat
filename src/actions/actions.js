export const saveAccessToken = token => ({
  type: 'SAVE_ACCESS_TOKEN',
  token
})

export const clearAccessToken = () => ({
  type: 'CLEAR_ACCESS_TOKEN'
})

export const logInUser = () => ({
  type: 'LOG_IN_USER',
  loggedIn: true
})

export const logOutUser = () => ({
  type: 'LOG_OUT_USER',
  loggedIn: false
})

export const saveUserName = userName => ({
  type: 'SAVE_USER_NAME',
  userName
})

export const selectSpm = spm => ({
  type: 'SELECT_SPM',
  spm
})

export const selectGenre = genre => ({
  type: 'SELECT_GENRE',
  genre
})

export const saveSpm = spm => ({
  type: 'SAVE_SPM',
  spm
})