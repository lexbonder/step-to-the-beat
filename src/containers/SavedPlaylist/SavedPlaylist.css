.saved-playlist {
  display: grid;
  grid-template-columns: auto 100px 65px;
  align-items: center;
  padding: 10px;
  border-bottom: 1px solid #151515;
}

.playlist-name {
  color: #ffffff;
  font-weight: normal;
  display: inline-block;
}

.playlist-details {
  color: #555555;
}

.delete {
  width: 47px;
  border-radius: 100%;
  color: red;
  border-color: red;
}

.delete {
  width: 47px;
  border-radius: 100%;
  color: #FF4A19;
  border-color: #FF4A19;
}

.delete:hover {
  color: #FF6C3B;
  border-color: #FF6C3B;
}

.delete:active {
  color: #DD4907;
  border-color: #DD4907
}