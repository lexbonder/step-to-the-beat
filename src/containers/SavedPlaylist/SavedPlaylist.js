import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { savePlaylist, selectSeed } from '../../actions/actions';
import { getPlaylistData } from '../../apiCalls';
import { playlistCleaner } from '../../playlistCleaner';
import PropTypes from 'prop-types';
import './SavedPlaylist.css';

export class SavedPlaylist extends Component {
  constructor(props) {
    super(props);
    this.state = {
      errorState: '',
      spm: '',
      genre: '',
      playlistName: ''
    };
  }

  componentDidMount = () => {
    const { user, seed } = this.props;
    const {spm, genre} = seed;
    this.setState({
      spm,
      genre,
      playlistName: `${user.name}'s ${spm} SPM, ${genre} playlist`
    });
  }

  handleClick = async () => {
    const {seed, accessToken, history } = this.props;
    const { spm, genre } = seed;
    try {
      const rawPlaylistData = await getPlaylistData(spm, genre, accessToken);
      const cleanedPlaylist = playlistCleaner(rawPlaylistData.tracks);
      this.props.savePlaylist(cleanedPlaylist);
      this.props.selectSeed(seed);
      history.push('/playlist');
    } catch (error) {
      this.setState({errorState: error.message});
    }
  }

  render() {
    const {playlistName, spm, genre} = this.state;
    return (
      <div className='saved-playlist'>
        <div>
          <h2 className='playlist-name'>
            {playlistName}
          </h2>
          <h3 className='playlist-details'>
            {spm} SPM &bull; {genre}
          </h3>
        </div>
        <button 
          className='next-and-back-buttons'
          onClick={this.handleClick}>
            View
        </button>
      </div>
    );
  }
}

const { string, func, shape, number } = PropTypes;

SavedPlaylist.propTypes = {
  user: shape({
    name: string,
    id: string,
    image: string
  }),
  accessToken: string,
  savePlaylist: func,
  selectSeed: func,
  seed: shape({
    spm: number,
    genre: string
  })
};

export const MSTP = ({accessToken, user}) => ({accessToken, user});

export const MDTP = dispatch => ({
  savePlaylist: playlist => dispatch(savePlaylist(playlist)),
  selectSeed: seed => dispatch(selectSeed(seed))
});

export default withRouter(connect(MSTP, MDTP)(SavedPlaylist));