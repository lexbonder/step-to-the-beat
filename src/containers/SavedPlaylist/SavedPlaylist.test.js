/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { SavedPlaylist, MSTP, MDTP } from './SavedPlaylist';
import { savePlaylist, selectSeed } from '../../actions/actions';

jest.mock('../../apiCalls')

describe('SavedPlaylist', () => {
  let wrapper;
  let mockSeed = {spm: 148, genre: 'ska'}
  let mockUser = {name: 'Alex', id: 'lxbndr', image: 'superlongurl'}
  let mockAccessToken = '12345abcde';
  let mockSavePlaylist = jest.fn()
  let mockSelectSeed = jest.fn()
  let mockHistory = { push: jest.fn() }

  beforeEach(() => {
    wrapper = shallow(<SavedPlaylist
      user={mockUser}
      seed={mockSeed} 
      accessToken={mockAccessToken}
      savePlaylist={mockSavePlaylist}
      selectSeed={mockSelectSeed}
      history={mockHistory} 
    />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('componentDidMount', () => {
    it('should set genre playlistName and spm after mounting', () => {
      const expectedState = {
        errorState: '',
        genre: 'ska',
        playlistName: `Alex's 148 SPM, ska playlist`,
        spm: 148,
      }
      expect(wrapper.state()).toEqual(expectedState)
    })
  })

  describe('handleClick', () => {
    it('should call savePlaylist with a cleaned playlist object', async () => {
      const mockCleanedPlaylist = [{
        artist: "I fight Dragons",
        id: "12341234",
        title: "Dont you?",
        uri: "spotify:track:12341234"
      }]
      await wrapper.instance().handleClick()
      expect(mockSavePlaylist).toHaveBeenCalledWith(mockCleanedPlaylist)
    })

    it('should call selectSeed with a seed object', async () => {
      await wrapper.instance().handleClick()
      expect(mockSelectSeed).toHaveBeenCalledWith(mockSeed)
    })

    it('should redirect the user to /playlist', async () => {
      await wrapper.instance().handleClick()
      expect(mockHistory.push).toHaveBeenCalledWith('/playlist')
    })

    it('should set an error message into state if the fetch is unsuccessful', async () => {
      await wrapper.instance().handleClick()
      await wrapper.instance().handleClick()
      expect(wrapper.state().errorState).not.toEqual('')
    })
  })

  describe('MSTP', () => {
    it('should return an object with the values it gets from state', () => {
      const mockState = {
        accessToken: mockAccessToken,
        user: {
          name: 'Alex',
          id: 'lxbndr',
          image: 'superlongurl'
        }
      }

      const mapped = MSTP(mockState)

      expect(mapped.accessToken).toEqual(mockAccessToken)
      expect(mapped.user).toEqual(mockUser)
    })
  })

  describe('MDTP', () => {
    let mockDispatch = jest.fn()
    let result;

    beforeEach(() => {
      result = MDTP(mockDispatch)
    })

    it('should call dispatch when savePlaylist is called', () => {
      const mockPlaylist = [{
        artist: "I fight Dragons",
        id: "12341234",
        title: "Dont you?",
        uri: "spotify:track:12341234"
      }]

      result.savePlaylist(mockPlaylist)
      expect(mockDispatch).toHaveBeenCalledWith(savePlaylist(mockPlaylist))
    })

    it('should call dispatch when selectSeed is called', () => {
      result.selectSeed(mockSeed)
      expect(mockDispatch).toHaveBeenCalledWith(selectSeed(mockSeed))
    })
  })
})