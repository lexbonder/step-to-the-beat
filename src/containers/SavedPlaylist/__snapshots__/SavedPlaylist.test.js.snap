// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedPlaylist should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SavedPlaylist
    accessToken="12345abcde"
    deleteSeed={[Function]}
    history={
        Object {
            "push": [Function],
          }
    }
    savePlaylist={[Function]}
    seed={
        Object {
            "genre": "ska",
            "id": 123,
            "spm": 148,
          }
    }
    selectSeed={[Function]}
    user={
        Object {
            "id": "lxbndr",
            "image": "superlongurl",
            "name": "Alex",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <h2
                    className="playlist-name"
          >
                    Alex's 148 SPM, ska playlist
          </h2>
          <h3
                    className="playlist-details"
          >
                    148
                     SPM • 
                    ska
          </h3>
</div>,
        <button
          className="buttons"
          onClick={[Function]}
>
          View
</button>,
        <button
          className="buttons delete"
          id={123}
          onClick={[Function]}
>
          X
</button>,
      ],
      "className": "saved-playlist",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2
              className="playlist-name"
>
              Alex's 148 SPM, ska playlist
</h2>,
            <h3
              className="playlist-details"
>
              148
               SPM • 
              ska
</h3>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Alex's 148 SPM, ska playlist",
              "className": "playlist-name",
            },
            "ref": null,
            "rendered": "Alex's 148 SPM, ska playlist",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                148,
                " SPM • ",
                "ska",
              ],
              "className": "playlist-details",
            },
            "ref": null,
            "rendered": Array [
              148,
              " SPM • ",
              "ska",
            ],
            "type": "h3",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "View",
          "className": "buttons",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "View",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "X",
          "className": "buttons delete",
          "id": 123,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "X",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <h2
                        className="playlist-name"
            >
                        Alex's 148 SPM, ska playlist
            </h2>
            <h3
                        className="playlist-details"
            >
                        148
                         SPM • 
                        ska
            </h3>
</div>,
          <button
            className="buttons"
            onClick={[Function]}
>
            View
</button>,
          <button
            className="buttons delete"
            id={123}
            onClick={[Function]}
>
            X
</button>,
        ],
        "className": "saved-playlist",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2
                className="playlist-name"
>
                Alex's 148 SPM, ska playlist
</h2>,
              <h3
                className="playlist-details"
>
                148
                 SPM • 
                ska
</h3>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Alex's 148 SPM, ska playlist",
                "className": "playlist-name",
              },
              "ref": null,
              "rendered": "Alex's 148 SPM, ska playlist",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  148,
                  " SPM • ",
                  "ska",
                ],
                "className": "playlist-details",
              },
              "ref": null,
              "rendered": Array [
                148,
                " SPM • ",
                "ska",
              ],
              "type": "h3",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "View",
            "className": "buttons",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "View",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "X",
            "className": "buttons delete",
            "id": 123,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "X",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
